design/external-authorization-design.md

```diff
@@ -0,0 +1,453 @@
+Configuring a single authorization service for all the virtual hosts might be acceptable for some organizations, but it doesn't really fix the Contour multi-tenancy model.
+On the other hand, Contour configures HTTPS with a separate HTTP Connection Manager for each virtual host, so different authorization servers can naturally be attached to each HTTPS virtual host.
+
+The result of this is that authorization servers are not supported on HTTP.
+While it is convenient from an implementation perspective, this policy is also consistent with Contour's security-first posture.
+Note that using the TLS fallback certificate (for non-SNI clients) has the same HTTP Connection Manager properties as HTTP, so authentication servers also cannot be configured on virtual hosts that share the fallback certificate.
+
+## Detailed Design
+
+### SupportService Changes
+
+The `SupportService` CRD has API version `projectcontour.io/v1alpha1`.
+
+```Go
+// SupportProtocolVersion is the version of the GRPC protocol used
+// to request support services.
+type SupportProtocolVersion string
+
+// SupportProtocolVersion2 requests the "v2" support protocol version.
+const SupportProtocolVersion2 SupportProtocolVersion = "v2"
+
+type SupportService struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   SupportServiceSpec   `json:"spec,omitempty"`
+	Status SupportServiceStatus `json:"status,omitempty"`
+}
+
+type SupportServiceStatus struct {
+	Conditions []Condition `json:"conditions"`
+}
+
+type SupportServiceSpec struct {
+	// +required
+	Service contourv1.Service `json:"service"`
+
+	// The load balancing policy for sending requests to the service.
+	// +optional
+	LoadBalancerPolicy *contourv1.LoadBalancerPolicy `json:"loadBalancerPolicy,omitempty"`
+
+	// This field sets the version of the GRPC protocol that Envoy uses to
+	// send requests to the support service. Since Contour always uses the
+	// v2 Envoy API, this is currently fixed at "v2". However, other
+	// protocol options will be available in future.
+	//
+	// +kubebuilder:validation:Enum=v2
+	ProtocolVersion SupportProtocolVersion `json:"protocolVersion,omitempty"`
+}
+```
+
+The `Conditions` field in `SupportServicesStatus` follows the standardized upstream structure.
+This type should be shared with other Contour API status types.
+
+This document does not yet define how Contour monitors the support service.
+However, the presence of the `Conditions` field allows Contour to expose the status of the support service fairly directly.
+At minimum, Contour should monitor the underlying Kubernetes service and expose whether there are sufficient healthy endpoints.
+
+Each `SupportService` CRD generates a unique upstream Envoy [Cluster][#4], which will emit standard Envoy metrics.
+Note that the Envoy cluster name can be non-obvious, so exposing it in status may be helpful.
+
+The `SupportService` CRD re-uses the `Service` type from the `projectcontour.io/v1` API.
```

Comment:
I don't think I'm keen on this idea of re-use. If we're going to build a new CRD, let's do that and make the user experience simple.

Reply:
In re-using the types, my goal was to keep it simple, since users should already be familiar with the service struct from using the HTTPProxy.

Reply:
We got burned by this before when re-using. It's an internal type, I'd rather bite the work to create a 5 line struct vs all the additional logic to make sure users don't reuse this field incorrectly due to the different context.

Reply:
OK, I can copy a matching Service type across, leaving out the unnecessary fields. I guess it will need to have a unique name, since this CRD will eventually migrate to the same package.

Reply:
I think it's clearer to have a similar type that doesn't have restrictions about not setting fields that are present - better to just not have fields that aren't relevant.

Reply:
I looked at this, and we would also have to duplicate the `UpstreamValidation`, `HeadersPolicy` and `LoadBalancerPolicy`. That's possible, but it's also pretty intrusive; it means that we will have to do substantial refactoring to be able to apply helpers like `determineSNI()`, `getProtocol()`, etc. However, it's also possible that this refactoring will end up with cleaner internal APIs (maybe create a `model` package for types that compose the DAG?).

We also need to keep in mind that this API will graduate to `projectcontour/v1` and make sure that what we do now will make sense in terms of that destination. Would we want to deduplicate the types at that point, or keep the more specific types?

Reply:
Really? That much duplication? We can't inline the policies? I only meant that it's better not to use something that has fields that you shouldn't set - that's a pretty confusing UX.

I also hadn't though about methods and type-passing, that will make copying the type a significant refactor. I'm not so enamoured of the idea with that much extra work added.

Reply:
I posted in slack, but why do you need UpstreamValidation, HeadersPolicy and LoadBalancerPolicy? We're just connecting Envoy to another service. That piece is confusing to me which is why I suggested not to resue the service struct.

Reply:
Envoy should connect to the service over TLS, so you need the `UpstreamValidationPolicy` to express how to check the server certificate. If the service is an externalName service, you might need to rewrite the host header, so you'd need `HeadersPolicy`. There could be other operational reasons to add the headers, e.g. assigning Linkerd headers. The service probably has multiple pods, so `LoadBalancerPolicy` is probably a useful control, otherwise we would just hard-code a load balancing policy.

Reply:
Those are the bits I don't understand. 

- How would an external name service fit? The goal of this CRD is to connect Envoy to the Auth service over gRPC. Are you saying folks would deploy the Auth server outside the cluster and Envoy would reach out over the external name service? 
- Kubernetes doesn't provide a way to set load balancer policies, is that really required for this implementation?
- Is external name the only reason to have a Headers Policy? This is another field I don't understand how it would be applied in this situation and be useful. 

It seems like the only reason to  have the new crd is to allow for setting conditions, but I have questions about how that will work.

Reply:
Was your thought on headers policy to map to the different response types like `OkHttpResponse`?   https://www.envoyproxy.io/docs/envoy/v1.14.3/api-v2/service/auth/v2/external_auth.proto#service-auth-v2-deniedhttpresponse

Reply:
> Those are the bits I don't understand.
> 
> * How would an external name service fit? The goal of this CRD is to connect Envoy to the Auth service over gRPC. Are you saying folks would deploy the Auth server outside the cluster and Envoy would reach out over the external name service?

Yes, this would work.

> * Kubernetes doesn't provide a way to set load balancer policies, is that really required for this implementation?

Required? Maybe not. But it's obviously useful and easy to implement.

> * Is external name the only reason to have a Headers Policy? This is another field I don't understand how it would be applied in this situation and be useful.

I gave 2 possible examples above. Probably creative operators can think of more uses for this.

> It seems like the only reason to have the new crd is to allow for setting conditions, but I have questions about how that will work.

OK?

> Was your thought on headers policy to map to the different response types like OkHttpResponse?

No, that's part of the GRPC payload; there's no way to intercept that.